File: packages/collector-go/internal/watcher/discovery.go
package watcher

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// AgentLogLocations defines default log paths per OS and agent
var AgentLogLocations = map[string]map[string][]string{
	"copilot": {
		"darwin": {
			"~/Library/Application Support/Code/User/workspaceStorage/*/chatSessions",
			"~/Library/Application Support/Code - Insiders/User/workspaceStorage/*/chatSessions",
		},
		"linux": {
			"~/.config/Code/User/workspaceStorage/*/chatSessions",
			"~/.config/Code - Insiders/User/workspaceStorage/*/chatSessions",
		},
		"windows": {
			"%APPDATA%\\Code\\User\\workspaceStorage\\*\\chatSessions",
			"%APPDATA%\\Code - Insiders\\User\\workspaceStorage\\*\\chatSessions",
		},
	},
	"claude": {
		"darwin": {
			"~/.claude/logs",
			"~/Library/Application Support/Claude/logs",
			"~/Library/Logs/Claude",
		},
		"linux": {
			"~/.claude/logs",
			"~/.config/claude/logs",
			"~/.local/share/claude/logs",
		},
		"windows": {
			"%APPDATA%\\Claude\\logs",
			"%LOCALAPPDATA%\\Claude\\logs",
		},
	},
	"cursor": {
		"darwin": {
			"~/Library/Application Support/Cursor/logs",
			"~/Library/Logs/Cursor",
		},
		"linux": {
			"~/.config/Cursor/logs",
			"~/.local/share/Cursor/logs",
		},
		"windows": {
			"%APPDATA%\\Cursor\\logs",
			"%LOCALAPPDATA%\\Cursor\\logs",
		},
	},
	"cline": {
		"darwin": {
			"~/.vscode/extensions/saoudrizwan.claude-dev-*/logs",
			"~/Library/Application Support/Code/logs/*/exthost",
		},
		"linux": {
			"~/.vscode/extensions/saoudrizwan.claude-dev-*/logs",
			"~/.config/Code/logs/*/exthost",
		},
		"windows": {
			"%USERPROFILE%\\.vscode\\extensions\\saoudrizwan.claude-dev-*\\logs",
			"%APPDATA%\\Code\\logs\\*\\exthost",
		},
	},
	"aider": {
		"darwin": {
			"~/.aider/logs",
			"~/.aider/.aider.history",
		},
		"linux": {
			"~/.aider/logs",
			"~/.aider/.aider.history",
		},
		"windows": {
			"%USERPROFILE%\\.aider\\logs",
			"%USERPROFILE%\\.aider\\.aider.history",
		},
	},
}

// DiscoveredLog represents a discovered log file or directory
type DiscoveredLog struct {
	AgentName string
	Path      string
	IsDir     bool
	Exists    bool
}

// DiscoverAgentLogs finds actual log file locations for a specific agent
func DiscoverAgentLogs(agentName string) ([]DiscoveredLog, error) {
	osName := runtime.GOOS
	patterns, exists := AgentLogLocations[agentName]
	if !exists {
		return nil, fmt.Errorf("unknown agent: %s", agentName)
	}

	osPlatterns, exists := patterns[osName]
	if !exists {
		return nil, fmt.Errorf("agent %s not supported on %s", agentName, osName)
	}

	var discovered []DiscoveredLog

	for _, pattern := range osPlatterns {
		// Expand path variables
		expanded := expandPath(pattern)

		// Handle glob patterns
		matches, err := filepath.Glob(expanded)
		if err != nil {
			// Log error but continue with other patterns
			continue
		}

		// Check each match
		for _, match := range matches {
			info, err := os.Stat(match)
			if err != nil {
				continue
			}

			discovered = append(discovered, DiscoveredLog{
				AgentName: agentName,
				Path:      match,
				IsDir:     info.IsDir(),
				Exists:    true,
			})
		}
	}

	return discovered, nil
}

// DiscoverAllAgentLogs discovers logs for all known agents
func DiscoverAllAgentLogs() (map[string][]DiscoveredLog, error) {
	result := make(map[string][]DiscoveredLog)

	for agentName := range AgentLogLocations {
		logs, err := DiscoverAgentLogs(agentName)
		if err != nil {
			continue // Skip agents that aren't supported on this OS
		}

		if len(logs) > 0 {
			result[agentName] = logs
		}
	}

	return result, nil
}

// FindLogFiles recursively finds log files in a directory
func FindLogFiles(dirPath string) ([]string, error) {
	var logFiles []string

	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // Continue on errors
		}

		if info.IsDir() {
			return nil
		}

		// Check if file is a log file
		if isLogFile(path) {
			logFiles = append(logFiles, path)
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	return logFiles, nil
}

// isLogFile checks if a file is likely a log file
func isLogFile(path string) bool {
	ext := strings.ToLower(filepath.Ext(path))
	base := strings.ToLower(filepath.Base(path))

	// Check common log file extensions
	logExtensions := []string{".log", ".txt", ".json", ".jsonl", ".ndjson"}
	for _, logExt := range logExtensions {
		if ext == logExt {
			return true
		}
	}

	// Check common log file patterns
	logPatterns := []string{
		"log",
		"output",
		"console",
		"trace",
		"debug",
		"error",
		"access",
	}
	for _, pattern := range logPatterns {
		if strings.Contains(base, pattern) {
			return true
		}
	}

	return false
}

// expandPath expands ~ and environment variables in a path
func expandPath(path string) string {
	// Expand ~
	if strings.HasPrefix(path, "~/") {
		homeDir, err := os.UserHomeDir()
		if err == nil {
			path = filepath.Join(homeDir, path[2:])
		}
	}

	// Expand environment variables
	if runtime.GOOS == "windows" {
		// Windows uses %VAR% syntax
		path = os.ExpandEnv(path)
	} else {
		// Unix uses $VAR or ${VAR} syntax
		path = os.ExpandEnv(path)
	}

	return path
}

// GetLatestLogFile finds the most recently modified log file in a directory
func GetLatestLogFile(dirPath string) (string, error) {
	files, err := FindLogFiles(dirPath)
	if err != nil {
		return "", err
	}

	if len(files) == 0 {
		return "", fmt.Errorf("no log files found in %s", dirPath)
	}

	// Find the most recent file
	var latestFile string
	var latestTime int64

	for _, file := range files {
		info, err := os.Stat(file)
		if err != nil {
			continue
		}

		modTime := info.ModTime().Unix()
		if modTime > latestTime {
			latestTime = modTime
			latestFile = file
		}
	}

	if latestFile == "" {
		return "", fmt.Errorf("no accessible log files found")
	}

	return latestFile, nil
}

File: rust/devlog-watcher/src/discovery.rs
use std::path::{Path, PathBuf};
use std::collections::HashMap;
use anyhow::{Result, anyhow};
use glob::glob;

pub struct DiscoveredLog {
    pub agent_name: String,
    pub path: PathBuf,
    pub is_dir: bool,
    pub exists: bool,
}

pub fn get_agent_log_locations() -> HashMap<String, HashMap<String, Vec<String>>> {
    let mut locations = HashMap::new();

    // Copilot
    let mut copilot = HashMap::new();
    copilot.insert("macos".to_string(), vec![
        "~/Library/Application Support/Code/User/workspaceStorage/*/chatSessions".to_string(),
        "~/Library/Application Support/Code - Insiders/User/workspaceStorage/*/chatSessions".to_string(),
    ]);
    copilot.insert("linux".to_string(), vec![
        "~/.config/Code/User/workspaceStorage/*/chatSessions".to_string(),
        "~/.config/Code - Insiders/User/workspaceStorage/*/chatSessions".to_string(),
    ]);
    copilot.insert("windows".to_string(), vec![
        "%APPDATA%\\Code\\User\\workspaceStorage\\*\\chatSessions".to_string(),
        "%APPDATA%\\Code - Insiders\\User\\workspaceStorage\\*\\chatSessions".to_string(),
    ]);
    locations.insert("copilot".to_string(), copilot);

    // Claude
    let mut claude = HashMap::new();
    claude.insert("macos".to_string(), vec![
        "~/.claude/logs".to_string(),
        "~/Library/Application Support/Claude/logs".to_string(),
        "~/Library/Logs/Claude".to_string(),
    ]);
    claude.insert("linux".to_string(), vec![
        "~/.claude/logs".to_string(),
        "~/.config/claude/logs".to_string(),
        "~/.local/share/claude/logs".to_string(),
    ]);
    claude.insert("windows".to_string(), vec![
        "%APPDATA%\\Claude\\logs".to_string(),
        "%LOCALAPPDATA%\\Claude\\logs".to_string(),
    ]);
    locations.insert("claude".to_string(), claude);

    // Add others as needed...
    locations
}

pub fn discover_agent_logs(agent_name: &str) -> Result<Vec<DiscoveredLog>> {
    let os_name = if cfg!(target_os = "macos") {
        "macos"
    } else if cfg!(target_os = "windows") {
        "windows"
    } else {
        "linux"
    };

    let locations = get_agent_log_locations();
    let agent_locations = locations.get(agent_name).ok_or_else(|| anyhow!("unknown agent: {}", agent_name))?;
    let patterns = agent_locations.get(os_name).ok_or_else(|| anyhow!("agent {} not supported on {}", agent_name, os_name))?;

    let mut discovered = Vec::new();

    for pattern in patterns {
        let expanded = expand_path(pattern);
        if let Ok(entries) = glob(&expanded) {
            for entry in entries {
                if let Ok(path) = entry {
                    if let Ok(metadata) = std::fs::metadata(&path) {
                        discovered.push(DiscoveredLog {
                            agent_name: agent_name.to_string(),
                            path,
                            is_dir: metadata.is_dir(),
                            exists: true,
                        });
                    }
                }
            }
        }
    }

    Ok(discovered)
}

pub fn discover_all_agent_logs() -> Result<HashMap<String, Vec<DiscoveredLog>>> {
    let mut result = HashMap::new();
    let locations = get_agent_log_locations();

    for agent_name in locations.keys() {
        if let Ok(logs) = discover_agent_logs(agent_name) {
            if !logs.is_empty() {
                result.insert(agent_name.clone(), logs);
            }
        }
    }

    Ok(result)
}

pub fn expand_path(path: &str) -> String {
    let mut expanded = path.to_string();

    // Expand ~
    if expanded.starts_with("~/") {
        if let Some(home_dir) = dirs::home_dir() {
            expanded = expanded.replacen("~/", &format!("{}/", home_dir.display()), 1);
        }
    }

    // Expand environment variables
    if cfg!(target_os = "windows") {
        // Simple %VAR% expansion
        for (key, value) in std::env::vars() {
            let pattern = format!("%{}%", key);
            expanded = expanded.replace(&pattern, &value);
        }
    } else {
        // Simple $VAR expansion
        for (key, value) in std::env::vars() {
            let pattern = format!("${}", key);
            expanded = expanded.replace(&pattern, &value);
            let pattern_braces = format!("${{{}}}", key);
            expanded = expanded.replace(&pattern_braces, &value);
        }
    }

    expanded
}

pub fn is_log_file(path: &Path) -> bool {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("").to_lowercase();
    let base = path.file_name().and_then(|n| n.to_str()).unwrap_or("").to_lowercase();

    let log_extensions = ["log", "txt", "json", "jsonl", "ndjson"];
    if log_extensions.contains(&ext.as_str()) {
        return true;
    }

    let log_patterns = ["log", "output", "console", "trace", "debug", "error", "access"];
    for pattern in log_patterns {
        if base.contains(pattern) {
            return true;
        }
    }

    false
}
